### PR TITLE
[WIP] Adding support for drawing tools

### PR DESCRIPTION
**Pull request by pgiraud**
Features are added to the feature overlay, and are listed in the mymaps bar.

### CODE REVIEW

File: geoportailv3/static/js/draw/drawdirective.js [hunk 1]
```diff
@@ -0,0 +1,149 @@
+/**
+ * @fileoverview This file provides a draw directive. This directive is used
+ * to create a draw panel in the page.
+ *
+ * Example:
+ *
+ * <app-draw app-draw-map="::mainCtrl.map"></app-draw>
```
Comment (elemoine) [inline geoportailv3/static/js/draw/drawdirective.js]: Example is incomplete.

File: geoportailv3/static/js/draw/drawdirective.js [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+};
+
+
+/**
+ * @param {ol.interaction.Draw} interaction
+ * @param {ol.interaction.DrawEvent} event
+ * @private
+ */
+app.DrawController.prototype.onDrawEnd_ = function(interaction, event) {
+  event.feature.set('name', 'element ' + (++this.featureSeq_));
```
Comment (elemoine) [inline geoportailv3/static/js/draw/drawdirective.js]: Is this actuall used?

File: geoportailv3/static/js/draw/drawdirective.js [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+   * @type {angular.Scope}
+   * @private
+   */
+  this.scope_ = $scope;
+
+  var drawPoint = new ol.interaction.Draw({
+    features: this.features,
+    type: /** @type {ol.geom.GeometryType} */ ('Point')
+  });
+  /**
```
Comment (elemoine) [inline geoportailv3/static/js/draw/drawdirective.js]: Please insert a blank line before this line.

File: geoportailv3/static/js/draw/drawdirective.js [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+  /**
+   * @type {angular.Scope}
+   * @private
+   */
+  this.scope_ = $scope;
+
+  var drawPoint = new ol.interaction.Draw({
+    features: this.features,
+    type: /** @type {ol.geom.GeometryType} */ ('Point')
```
Comment (elemoine) [inline geoportailv3/static/js/draw/drawdirective.js]: You don't need a cast here in compiled code. Use `ol.geom.GeometryPoint.POINT`. And do not forget to add `goog.require('ol.geom.GeometryPoint')`.

File: geoportailv3/static/js/draw/drawdirective.js [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+  drawPoint.setActive(false);
+  ngeoDecorateInteraction(drawPoint);
+  this.map.addInteraction(drawPoint);
+  goog.events.listen(drawPoint, ol.interaction.DrawEventType.DRAWEND,
+      angular.bind(this, this.onDrawEnd_, drawPoint));
+
+  var drawLine = new ol.interaction.Draw({
+    features: this.features,
+    type: /** @type {ol.geom.GeometryType} */ ('LineString')
```
Comment (elemoine) [inline geoportailv3/static/js/draw/drawdirective.js]: Use `ol.geom.GeometryType.LINE_STRING`.

File: geoportailv3/static/js/draw/drawdirective.js [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+   * @type {ol.interaction.Draw}
+   * @export
+   */
+  this.drawPoint = drawPoint;
+
+  drawPoint.setActive(false);
+  ngeoDecorateInteraction(drawPoint);
+  this.map.addInteraction(drawPoint);
+  goog.events.listen(drawPoint, ol.interaction.DrawEventType.DRAWEND,
+      angular.bind(this, this.onDrawEnd_, drawPoint));
```
Comment (elemoine) [inline geoportailv3/static/js/draw/drawdirective.js]: Don't use `angular.bind` in compiled code. And you don't need to use `bind` at all here. Instead you can just use the following:

```js
  goog.events.listen(drawPoint, ol.interaction.DrawEventType.DRAWEND,
      this.onDrawEnd_, false, this);
```

and use `event.target` in `onDrawEnd_` to get a reference to the interaction to which the `drawend` event was dispatched.

File: geoportailv3/static/js/draw/drawdirective.js [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+   * @type {ol.interaction.Draw}
+   * @export
+   */
+  this.drawLine = drawLine;
+
+  drawLine.setActive(false);
+  ngeoDecorateInteraction(drawLine);
+  this.map.addInteraction(drawLine);
+  goog.events.listen(drawLine, ol.interaction.DrawEventType.DRAWEND,
+      angular.bind(this, this.onDrawEnd_, drawLine));
```
Comment (elemoine) [inline geoportailv3/static/js/draw/drawdirective.js]: Same comment as before for `bind`.

File: geoportailv3/static/js/draw/drawdirective.js [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+ * @param {ol.interaction.DrawEvent} event
+ * @private
+ */
+app.DrawController.prototype.onDrawEnd_ = function(interaction, event) {
+  event.feature.set('name', 'element ' + (++this.featureSeq_));
+
+  // Deactivating asynchronosly to prevent dbl-click to zoom in
+  window.setTimeout(goog.bind(function() {
+    interaction.setActive(false);
+    this.scope_.$applyAsync();
```
Comment (elemoine) [inline geoportailv3/static/js/draw/drawdirective.js]: It's a better practice to do the action in a function passed to `$apply/$applyAsync`:

```js
    this.scope_.$apply(function() {
      interection.setActive(false);
    });
```

And I don't think you need to use `$applyAsync` here, you can probably just use `$apply`.

File: geoportailv3/static/js/maincontroller.js [hunk 9]
```diff
@@ -211,6 +219,30 @@ app.MainController = function(
...
+  // synchronize the drawn features collection with the feature overlay
+  var drawOverlay = ngeoFeatureOverlayMgr.getFeatureOverlay();
+  $scope.$watchCollection(goog.bind(function() {
+    return this.drawnFeatures.getArray();
+  }, this), goog.bind(function() {
+    drawOverlay.clear();
+    this.drawnFeatures.forEach(function(feature) {
+      drawOverlay.addFeature(feature);
+    });
+  }, this));
```
Comment (elemoine) [inline geoportailv3/static/js/maincontroller.js]: As discussed I think the synchronization could/should be done by the feature overlay.

File: geoportailv3/static/js/mymaps/mymapsdirective.js [hunk 10]
```diff
@@ -101,6 +110,18 @@ app.MymapsDirectiveController = function($scope, $compile, gettext) {
    * @export
    */
   this.newDescription = '';
+
+  /**
+   * @type {Array.<ol.Feature>}
+   * @export
+   */
+  this.featuresList = this['features'].getArray();
```
Comment (elemoine) [inline geoportailv3/static/js/mymaps/mymapsdirective.js]: You should declare `this.features` with an `@export`. In this way you won't need to use the square bracket notation.

File: geoportailv3/static/js/mymaps/mymapsdirective.js [hunk 11]
```diff
@@ -45,6 +48,12 @@ app.module.directive('appMymaps', app.mymapsDirective);
 app.MymapsDirectiveController = function($scope, $compile, gettext) {
 
   /**
+   * @type {ol.interaction.Select}
+   *  @export
```
Comment (elemoine) [inline geoportailv3/static/js/mymaps/mymapsdirective.js]: Do you really need to export `selectInteraction`? (I don't think you refer to `selectInteraction` from a template.)

File: geoportailv3/templates/index.html [hunk 12]
```diff
@@ -145,7 +145,9 @@ <h2 translate>my_maps</h2>
             ✕
           </button>
           <div class="col-xs-12">
-            <app-mymaps><app-mymaps>
+            <app-mymaps app-mymaps-selectinteraction="mainCtrl.selectInteraction"
+              app-mymaps-features="mainCtrl.drawnFeatures">
```
Comment (elemoine) [inline geoportailv3/templates/index.html]: You should use one-time binding for both `selectInteraction` and `drawnFeatures`.

File: geoportailv3/templates/index.html [hunk 13]
```diff
@@ -206,13 +208,11 @@ <h2 translate>infos</h2>
...
-          <div class="drawing-panel">
-              <div class="col-xs-12">
-              <p translate>The drawing tools are currently not available in this preview</p>
-            </div>
-          </div>
-        </div>
+          <app-draw app-draw-map="::mainCtrl.map"
+            app-draw-active="mainCtrl.drawOpen"
+            app-draw-selectinteraction="mainCtrl.selectInteraction"
+            app-draw-features="mainCtrl.drawnFeatures">
```
Comment (elemoine) [inline geoportailv3/templates/index.html]: Use one-time binding for both `selectInteraction` and `features`.

**Comment by elemoine** [comment]
Firt review round finished.

File: geoportailv3/static/js/draw/drawdirective.js (hunk 2 above)
Comment (pgiraud) [inline geoportailv3/static/js/draw/drawdirective.js]: Yes, the name is used to be displayed in the mymaps list.
However, I expect it to be a bit more complex to get the geometry type used as well (ie. "point 0", "line 1", etc...)

File: geoportailv3/static/js/mymaps/mymapsdirective.js (hunk 11 above)
Comment (pgiraud) [inline geoportailv3/static/js/mymaps/mymapsdirective.js]: As far as I understand, if I don't export it, I won't be able to use the dot notation.

File: geoportailv3/static/js/maincontroller.js [hunk 14]
```diff
@@ -211,6 +219,30 @@ app.MainController = function(
       $('app-themeswitcher #themes-content').collapse('hide');
     }
   }, this));
+
+  var selectInteraction = new ol.interaction.Select({
+    filter: goog.bind(function(feature, layer) {
+      return this.drawnFeatures.getArray().indexOf(feature) != -1;
+    }, this)
```
Comment (elemoine) [inline geoportailv3/static/js/maincontroller.js]: As an optimization you can also add a layer filter:

```js
    layers: [ngeoFeatureOverlayMgr.getLayer()]
```

With this OpenLayers' hit detection mechanism will skip other layers.

Comment (pgiraud) [inline geoportailv3/static/js/maincontroller.js]: For an unknown reason this doesn't work. I wonder if this is a bug in OL3. I'll have to check.

**Comment by pgiraud** [comment]
I took last comments from Éric and fixed a small issue with selection not being taken into account synchronously.
@jaykayone Do you think we can merge this now in order to be able to continue on other topics? I can still be improved later.

**Comment by jaykayone** [comment]
yes, go ahead

File: geoportailv3/static/js/maincontroller.js (hunk 14 above)
Comment (elemoine) [inline geoportailv3/static/js/maincontroller.js]: > For an unknown reason this doesn't work. I wonder if this is a bug in OL3. I'll have to check.

Curious to know if you found anything.

Comment (pgiraud) [inline geoportailv3/static/js/maincontroller.js]: https://github.com/openlayers/ol3/issues/2940#issuecomment-140081090

Comment (elemoine) [inline geoportailv3/static/js/maincontroller.js]: I created a path for that. See https://github.com/openlayers/ol3/pull/4143. Let's see what the other developers say about it.

**Pull request merged**